File: java/patterns/src/main/java/org/prototypepatterns/structural/decorator/DecoratorPattern.java
package org.prototypepatterns.structural.decorator;

public final class DecoratorPattern {
    private DecoratorPattern() {
    }

    interface Notifier {
        String send(String message);
    }

    static final class EmailNotifier implements Notifier {
        @Override
        public String send(String message) {
            return "email:" + message;
        }
    }

    static final class SmsDecorator implements Notifier {
        private final Notifier wrapped;

        SmsDecorator(Notifier wrapped) {
            this.wrapped = wrapped;
        }

        @Override
        public String send(String message) {
            return wrapped.send(message) + "|sms:" + message;
        }
    }

    static final class SlackDecorator implements Notifier {
        private final Notifier wrapped;

        SlackDecorator(Notifier wrapped) {
            this.wrapped = wrapped;
        }

        @Override
        public String send(String message) {
            return wrapped.send(message) + "|slack:" + message;
        }
    }

    public static String demo() {
        Notifier notifier = new SlackDecorator(new SmsDecorator(new EmailNotifier()));
        return notifier.send("build complete");
    }
}

File: csharp/patterns/src/PrototypePatterns/Structural/Decorator/DecoratorPattern.cs
namespace PrototypePatterns.Structural.Decorator;

public interface INotifier
{
    string Send(string message);
}

public sealed class EmailNotifier : INotifier
{
    public string Send(string message) => $"email:{message}";
}

public sealed class SmsDecorator(INotifier wrapped) : INotifier
{
    public string Send(string message) => $"{wrapped.Send(message)}|sms:{message}";
}

public sealed class SlackDecorator(INotifier wrapped) : INotifier
{
    public string Send(string message) => $"{wrapped.Send(message)}|slack:{message}";
}

public static class DecoratorPattern
{
    public static string Demo() => new SlackDecorator(new SmsDecorator(new EmailNotifier())).Send("build complete");
}
